File: src/aside/java/org/taktik/icure/logic/impl/KeywordLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.dao.KeywordDAO;
import org.taktik.icure.dao.impl.idgenerators.UUIDGenerator;
import org.taktik.icure.entities.Keyword;
import org.taktik.icure.entities.embed.Delegation;
import org.taktik.icure.logic.KeywordLogic;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.validation.aspect.Check;

import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;


@org.springframework.stereotype.Service
public class KeywordLogicImpl extends GenericLogicImpl<Keyword, KeywordDAO> implements KeywordLogic {
	private static final Logger log = LoggerFactory.getLogger(KeywordLogicImpl.class);


	private KeywordDAO keywordDAO;
	private UUIDGenerator uuidGenerator;
	private ICureSessionLogic sessionLogic;

	@Autowired
	public void setKeywordDAO(KeywordDAO keywordDAO) {
		this.keywordDAO = keywordDAO;
	}
	@Autowired
	public void setUuidGenerator(UUIDGenerator uuidGenerator) {
		this.uuidGenerator = uuidGenerator;
	}
	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}
	@Override
	protected KeywordDAO getGenericDAO() {
		return keywordDAO;
	}

	@Override
	public Keyword createKeyword(@Check @NotNull Keyword keyword) {
		List<Keyword> createdKeywords = new ArrayList<>(1);
		try {
			// Setting Keyword attributes
			keyword.setId(uuidGenerator.newGUID().toString());
			createEntities(Collections.singleton(keyword), createdKeywords);
		} catch (Exception e) {
			log.error("createKeyword: " + e.getMessage());
			throw new IllegalArgumentException("Invalid Keyword", e);
		}
		return createdKeywords.size() == 0 ? null:createdKeywords.get(0);
	}

	@Override
	public Keyword getKeyword(String keywordId) {
		return keywordDAO.getKeyword(keywordId);
	}


	@Override
	public Set<String> deleteKeywords(Set<String> ids) {
		try {
			deleteEntities(ids);
		} catch (Exception e) {
			log.error(e.getMessage(), e);
			return null;
		}
		return ids;
	}

	@Override
	public Keyword modifyKeyword(@Check @NotNull Keyword keyword) {
		try {
			updateEntities(Collections.singleton(keyword));
			return getKeyword(keyword.getId());
		} catch (Exception e) {
			throw new IllegalArgumentException("Invalid Keyword", e);
		}
	}

	@Override
	public List<Keyword> getKeywordsByUser(String userId) {
		return keywordDAO.getByUserId(userId);
	}

}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/KeywordLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOf
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.taktik.couchdb.DocIdentifier
import org.taktik.icure.asyncdao.KeywordDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.KeywordLogic
import org.taktik.icure.dao.impl.idgenerators.UUIDGenerator
import org.taktik.icure.entities.Keyword
import org.taktik.icure.utils.firstOrNull

@ExperimentalCoroutinesApi
@Service
class KeywordLogicImpl(private val keywordDAO: KeywordDAO,
                       private val uuidGenerator: UUIDGenerator,
                       private val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<Keyword, KeywordDAO>(sessionLogic), KeywordLogic {

    override fun getGenericDAO(): KeywordDAO {
        return keywordDAO
    }

    override suspend fun createKeyword(keyword: Keyword) = fix(keyword) { keyword ->
        val createdKeywords = try { // Setting Keyword attributes
            createEntities(setOf(keyword))
        } catch (e: Exception) {
            log.error("createKeyword: " + e.message)
            throw IllegalArgumentException("Invalid Keyword", e)
        }
        createdKeywords.firstOrNull()
    }

    override suspend fun getKeyword(keywordId: String): Keyword? {
        return keywordDAO.getKeyword(keywordId)
    }

    override fun deleteKeywords(ids: Set<String>): Flow<DocIdentifier> {
        return try {
            deleteByIds(ids)
        } catch (e: Exception) {
            log.error(e.message, e)
            flowOf()
        }
    }

    override suspend fun modifyKeyword(keyword: Keyword): Keyword? {
        return try {
            updateEntities(setOf(keyword)).firstOrNull()
        } catch (e: Exception) {
            throw IllegalArgumentException("Invalid Keyword", e)
        }
    }

    override fun getKeywordsByUser(userId: String): Flow<Keyword> = flow {
        emitAll(keywordDAO.getByUserId(userId))
    }

    companion object {
        private val log = LoggerFactory.getLogger(KeywordLogicImpl::class.java)
    }
}
